kratos/sources/connectivity_preserve_modeler.cpp

```diff
@@ -154,19 +198,19 @@ void ConnectivityPreserveModeler::DuplicateCommunicatorData(
     Communicator::Pointer pDestinationComm = rReferenceComm.Create();
     pDestinationComm->SetNumberOfColors( rReferenceComm.GetNumberOfColors() );
     pDestinationComm->NeighbourIndices() = rReferenceComm.NeighbourIndices();
-    pDestinationComm->LocalMesh().SetNodes( rReferenceComm.LocalMesh().pNodes() );
-    pDestinationComm->InterfaceMesh().SetNodes( rReferenceComm.InterfaceMesh().pNodes() );
-    pDestinationComm->GhostMesh().SetNodes( rReferenceComm.GhostMesh().pNodes() );
-    for (unsigned int i = 0; i < rReferenceComm.GetNumberOfColors(); i++) {
-        pDestinationComm->pLocalMesh(i)->SetNodes( rReferenceComm.pLocalMesh(i)->pNodes() );
-        pDestinationComm->pInterfaceMesh(i)->SetNodes( rReferenceComm.pInterfaceMesh(i)->pNodes() );
-        pDestinationComm->pGhostMesh(i)->SetNodes( rReferenceComm.pGhostMesh(i)->pNodes() );
-    }
 
     // This is a dirty hack to detect if the communicator is a Communicator or an MPICommunicator
     // Note that downcasting would not work here because MPICommunicator is not compiled in non-MPI builds
-    const bool is_mpi = ( rOriginModelPart.pElements() == rReferenceComm.LocalMesh().pElements() );
+    const bool is_mpi = ( rOriginModelPart.pElements() != rReferenceComm.LocalMesh().pElements() );
```

Comment:
why not using 
~~~
GetDefaultCommunicator().IsDistributed() 
~~~

?
the current hack looks ...well..very hacky to me. And i would say that it easily fails since elements in principle belong to a single modelpart, so the that normally Elements() and LocalMesh().Elements() should coincide

Reply:
This relies on an implementation detail: for a serial communicator, the local mesh is a pointer to the same container as the mesh elements, for an mpi communicator the container is different (a copy) even if the elements inside are the same.

In any case, I agree that it is hacky, and the correct check is coming in #4745